### PR TITLE
Use `downcast_primitive_array` in arithmetic kernels

### LINKED ISSUE #2639: Use `downcast_primitive_array` in arithmetic kernels
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The recently added `downcast_primitive_array` could be a standard macro inside this crate to downcast primitive array for primitive types and perform common operation on them. It would be nicer if we can use it in arithmetic kernels too.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by viirya**
# Which issue does this PR close?

Closes #2639.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by viirya** [comment]
Did a benchmark run and no obvious difference found.

File: arrow/src/array/cast.rs [hunk 1]
```diff
@@ -225,6 +225,255 @@ macro_rules! downcast_primitive_array {
             $($p => $fallback,)*
         }
     };
+
+    (($values1:ident, $values2:ident) => $e:block $($p:pat => $fallback:expr $(,)*)*) => {
```
Comment (tustvold) [inline arrow/src/array/cast.rs:229]: This has an implicit assumption that $values1 and $values2 have the same type, not only is this potentially surprising as an API, but I think it changes the behaviour of the kernels which will now panic where previously they would work?

Adding `$values2.data_type()` to the match might work, but it still feels a bit confusing? :thinking:

I wonder if we could instead do something like this

```
downcast_primitive_array!(
      left => {
          let right = as_primitive_array(right);
          multiply(left, right).map(|a| Arc::new(a) as ArrayRef)
      }
      _ => Err(ArrowError::CastError(format!(
          "Unsupported data type {}, {}",
          left.data_type(), right.data_type()
      )))
  )
```

And rely on the fact the generic kernels constrain them to be the same type. I don't know, perhaps this is hack...

**Comment by tustvold** [comment]
Btw very cool to see these being used already :heart:

File: arrow/src/array/cast.rs (hunk 1 above)
Comment (viirya) [reply arrow/src/array/cast.rs:229]: Ah, good point. But I guess the suggested one will also panic on `let right = as_primitive_array(right);` if `right` is not same type?

Let me do a test.

Comment (viirya) [reply arrow/src/array/cast.rs:229]: Hmm, `math_op` doesn't constrain two sides of op should be same type. So
```rust
downcast_primitive_array!(
    left => {
        let right = as_primitive_array(right);
        math_op(left, right, |a, b| a + b).map(|a| Arc::new(a) as ArrayRef)
    }
    _ => Err(ArrowError::CastError(format!(
       "Unsupported data type {}, {}",
        left.data_type(), right.data_type()
   )))
)
```

will not constrain the right side type.

Comment (tustvold) [reply arrow/src/array/cast.rs:229]: > suggested one will

Yeah, you will still need to check the type, it is what it is

> doesn't constrain the type

That's why i suggested using the generic kernel not math_op directly 😃

Comment (viirya) [reply arrow/src/array/cast.rs:229]: Oh oh right. 😄 

Good news is it works.

But as `Float16Type` doesn't implement `ArrowNumericType`, I need to remove `Float16Type` pattern from `downcast_primitive_array!` to make it work.

For `simd` feature, seems f16 related APIs are not available so it appears not easy to let `Float16Type` implement `ArrowNumericType`.

Currently I leave single argument `downcast_primitive_array!` untouched and stick with two argument one.

**Review by tustvold — APPROVED** [review]

**Pull request merged**